src/ray/gcs/gcs_server/gcs_table_storage.cc

```diff
@@ -99,35 +108,64 @@ Status GcsTableWithJobId<Key, Data>::GetByJobId(const JobID &job_id,
     }
     callback(std::move(values));
   };
-  return this->store_client_->AsyncGetByIndex(
-      this->table_name_, job_id.Binary(), on_done);
+  return this->store_client_->AsyncMultiGet(this->table_name_, keys, on_done);
 }
 
 template <typename Key, typename Data>
 Status GcsTableWithJobId<Key, Data>::DeleteByJobId(const JobID &job_id,
                                                    const StatusCallback &callback) {
-  return this->store_client_->AsyncDeleteByIndex(
-      this->table_name_, job_id.Binary(), callback);
+  std::vector<Key> keys;
+  {
+    absl::MutexLock lock(&mutex_);
+    auto &key_set = index_[job_id];
+    for (auto &key : key_set) {
+      keys.push_back(key);
+    }
+  }
+  return BatchDelete(keys, callback);
 }
 
 template <typename Key, typename Data>
 Status GcsTableWithJobId<Key, Data>::Delete(const Key &key,
                                             const StatusCallback &callback) {
-  return this->store_client_->AsyncDeleteWithIndex(
-      this->table_name_, key.Binary(), GetJobIdFromKey(key).Binary(), callback);
+  return BatchDelete({key}, callback);
 }
 
 template <typename Key, typename Data>
 Status GcsTableWithJobId<Key, Data>::BatchDelete(const std::vector<Key> &keys,
                                                  const StatusCallback &callback) {
   std::vector<std::string> keys_to_delete;
-  std::vector<std::string> indexs_to_delete;
   for (auto key : keys) {
     keys_to_delete.push_back(key.Binary());
-    indexs_to_delete.push_back(GetJobIdFromKey(key).Binary());
   }
-  return this->store_client_->AsyncBatchDeleteWithIndex(
-      this->table_name_, keys_to_delete, indexs_to_delete, callback);
+  return this->store_client_->AsyncBatchDelete(
+      this->table_name_, keys_to_delete, [this, callback, keys](auto status) {
+        if (status.ok()) {
+          {
+            absl::MutexLock lock(&mutex_);
+            for (auto &key : keys) {
+              index_[GetJobIdFromKey(key)].erase(key);
+            }
+          }
+        }
+        if (callback) {
+          callback(status);
+        }
+      });
+}
+
+template <typename Key, typename Data>
+Status GcsTableWithJobId<Key, Data>::AsyncRebuildIndexAndGetAll(
```

Comment:
No strong opinion here. But I feel this API is doing two things and seems like exposes the initialization logic to the caller.